shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java

```diff
@@ -46,6 +51,19 @@
 public class PlatformViewsController implements PlatformViewsAccessibilityDelegate {
   private static final String TAG = "PlatformViewsController";
 
+  // These view types allow out-of-band drawing commands that don't notify the Android view
+  // hierarchy.
+  // To support these cases, Flutter hosts the embedded view in a VirtualDisplay,
+  // and binds the VirtualDisplay to a GL texture that is then composed by the engine.
+  // However, there are a few issues with Virtual Displays. For example, they don't fully support
+  // accessibility due to https://github.com/flutter/flutter/issues/29717,
+  // and keyboard interactions may have non-derterministic behavior.
+  // Views that issue out-of-band drawing commands that aren't included in this array are
+  // required to call `View#invalidate()` to notify Flutter about the update.
+  // This isn't ideal, but given all the other limitations it's a reasonable tradeoff.
+  // Related issue: https://github.com/flutter/flutter/issues/103630
+  private static Class[] VIEW_TYPES_REQUIRE_VD = {SurfaceView.class};
```

Comment:
These limitations seem pretty severe.

What are we solving with this? Just unbreaking mapbox? Are there lots of use cases for SurfaceView inside a platform view? Could we be doing something better here and just giving mapbox a GL context to work with?

What views need to call invalidate? Just TextureViews?

Reply:
Of course, if we did give plugins some way to directly talk to GL, we'd have to have them do some heavier lifting around a11y and input anyway...

Reply:
But rather than having SurfaceView or TextureView in a platform view, could we just give plugins a GL context to draw into?

Alternatively, could the plugin just use the `TextureRegistry` API instead? That should generally perform better than platform views - but again, I'm not sure I really understand all the requirements. For a WebRTC plugin, I'd imagine they want to treat that more like how exoplayer works rather than like a platform view though...

Reply:
I would actually love to see is an official API that allows to project a View into a GL Texture while the view is in the view hierarchy. Some challenges:
1. it must support embedded TextureViews like the Google Maps View.
2. it must support embedded SurfaceViews like mapbox or WebRTC views. 
3. WebView is a special view that does its own thing

Reply:
My suggestion here is not that Android somehow give us an API to do this, but that we ask the mapbox and webRTC plugin teams to refactor their plugins to do this.

Reply:
> But rather than having SurfaceView or TextureView in a platform view, could we just give plugins a GL context to draw into?

Yes, they can today. :) I'd be great if they can go down that route. 

Part of the issue is that plugins maintainers don't own the core component, so they are constrained by the public API of these components, but if they could build or adapt it for Flutter, they could maybe just build it using the Flutter framework directly, or Flutter's texture driven by an external raster thread, but use Flutter a11y API, and Flutter widgets. It's a spectrum of options, but it all boils down to how much they can afford.

Reply:
Adding this API back gives us a relatively large API surface to maintain that has some severe known issues. I'd like to think we're driving people away from this API and towards APIs that will actually perform well.

How long ago did we remove this API? How many versions of Flutter exist without it? I think that might be the last piece I'm missing for context here. If this was only recently dropped and we're trying to quickly add it back it makes more sense to me, but it'd probably still be good to add it back as deprecated or something.

Reply:
> If this was only recently dropped and we're trying to quickly add it back it makes more sense to m

Yes, that is exactly the case. It was removed in https://github.com/flutter/engine/pull/31468/files, and it's currently in Flutter 3.0.x.

Reply:
There are many projects requiring the VirtualDisplay functionality. It involves every map view plugin including Google's, video player surfaces and some plugins with camera views. 

Many are linked from: https://github.com/flutter/flutter/issues/103630

I would suggest pushing through this revert until a working alternative with documentation on use is created. Plugin maintainers would gladly switch to a new working API if it existed, but there's no alternative to push them on to. This API was removed in the 3.0.0 major release and everything has remained broken since them for all plugins that utilize Android SDKs with native GL surfaces.

Reply:
I'm not opposed to relanding this. I wanted to make sure I understand the context.

That said, the video_player package does _not_ use this functionality - it uses external textures, which are much more efficient. We should try to find ways to drive plugins in more performant directions - I think it's pretty likely that platform views are never going to be the right solutino for plugins that want to draw to native GL surfaces.

Reply:
How would you get an activity provided by a native Android SDK that includes a widget which is a GL surface over to the Android side? Plugin developers in these cases aren't drawing a GL surface themselves, they're utilizing an SDK that does it for them. It would be wonderful to be able to use these SDKs without having stutter and jank.

Could you point me towards some documentation on piping GL surfaces straight over? In the Mapbox case, the widget in question that is usually inserted in to an activity is called a MapView https://docs.mapbox.com/android/maps/examples/simple-map-view/ and I'm pretty sure that all other map SDKs provide something similar without providing direct access to the underlying GL surface.

Reply:
For MapBox, instead of using `MapView` I believe you'd use `MapSurface` (https://docs.mapbox.com/android/maps/api/10.6.0/mapbox-maps-android/com.mapbox.maps/-map-surface/). 

You may be limited by particular SDKs, but you can also file feature requests in those SDKs to provide access to things like this - if they support SurfaceView/TextureView they should generally support rendering to a Surface itself though.

Reply:
You can see how VideoPlayer does it here: https://github.com/flutter/plugins/blob/main/packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java which uses the texture registry instead of a platform view.

Reply:
Filed https://github.com/flutter-mapbox-gl/maps/issues/1098 for MapBox specifically.

Reply:
Yes. Ideally, the MapBox plugin uses Surface directly, and the Texture widget. There's a nice article about this: https://medium.com/@german_saprykin/opengl-with-texture-widget-f919743d25d9 

/cc @felix-mittermeier

Reply:
MapView uses SurfaceView and TextureView https://github.com/mapbox/mapbox-maps-android/blob/main/sdk/src/main/java/com/mapbox/maps/MapView.kt

Reply:
> MapView uses SurfaceView and TextureView https://github.com/mapbox/mapbox-maps-android/blob/main/sdk/src/main/java/com/mapbox/maps/MapView.kt

Right, it's not best option for flutter Olivia

Reply:
Would it be possible to pipe the entire activity including overlay widgets through using the texture registry?

Reply:
No.  Just the surface